src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1559 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ *    Copyright 2024 (c) Fraunhofer IOSB (Author: Noel Graf)
+ */
+
+#include "ua_server_internal.h"
+
+#ifdef UA_ENABLE_GDS_PUSHMANAGEMENT
+
+#define UA_SHA1_LENGTH 20
+#define CHECKACTIVESESSIONINTERVAL 10000 /* 10sec */
+
+typedef struct UA_FileContext {
+    LIST_ENTRY(UA_FileContext) listEntry;
+    UA_ByteString *file;
+    UA_UInt32 fileHandle;
+    UA_NodeId sessionId;
+    UA_UInt64 currentPos;
+    UA_Byte openFileMode;
+} UA_FileContext;
+
+typedef struct UA_FileInfo {
+    UA_UInt16 openCount;
+    UA_UtcTime lastUpdateTime;
+    LIST_HEAD(, UA_FileContext)fileContext;
+} UA_FileInfo;
+
+typedef struct UA_FileInfoContext {
+    UA_NodeId certificateGroupId;
+    UA_FileInfo fileInfo;
+} UA_FileInfoContext;
+
+static UA_StatusCode applyChangesToServer(UA_Server *server);
+
+/* There are currently only two certification groups that are supported */
+size_t fileInfoContextSize = 2;
+UA_FileInfoContext fileInfoContext[2];
```

Comment:
Global variables should be const.
Mutable variables should go to the server struct.
So that we can have multiple server instances in the same application.

Either put it directly into the server struct.
Or define a "GDS manager struct" which aggregates them.